### PR TITLE
[REVIEW] add functions to set/get default CUDA stream

### PR DESCRIPTION
**Pull request by rongou**
First step for #6111.

@harrism @jrhemstad @kkraus14 @abellina

### CODE REVIEW

File: cpp/include/cudf/utilities/default_stream.hpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+               "Default stream can only be set to legacy or per-thread");
+  detail::default_stream = stream;
+}
+
+/**
+ * @brief Get the default CUDA stream.
+ *
+ * @returns cudaStream_t The default CUDA stream.
+ */
+cudaStream_t get_default_stream() { return detail::default_stream; }
```
Comment (harrism) [inline cpp/include/cudf/utilities/default_stream.hpp:52]: I think this should be made thread safe, and probably use a myers singleton like we had for `get_default_resource` (replaced by get_current_device_resource).

Also, this isn't safe for multidevice -- a user could set a non-default stream associated with device 0, which would then cause errors to use on device 1. So it probably should be `per_device_stream`...

**Review by harrism — CHANGES_REQUESTED** [review]

**Review by jrhemstad — CHANGES_REQUESTED** [review]
I don't think we should go about introducing this until we've landed and incorporated this work: https://github.com/rapidsai/rmm/pull/413

**Comment by harrism** [comment]
> I don't think we should go about introducing this until we've landed and incorporated this work: [rapidsai/rmm#413](https://github.com/rapidsai/rmm/pull/413)

I was planning to jump on that next. But at this point it's probably 0.17.

File: cpp/include/cudf/utilities/default_stream.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+#include <cuda_runtime_api.h>
+
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+
+namespace detail {
+#if __cpp_inline_variables
+inline cudaStream_t default_stream;
```
Comment (nvdbaranec) [inline cpp/include/cudf/utilities/default_stream.hpp:27]:
```suggestion
inline cudaStream_t default_stream{cudaStreamDefault};
```

**Review by nvdbaranec — CHANGES_REQUESTED** [review]

**Review by raydouglass — APPROVED** [review]

File: cpp/include/cudf/utilities/default_stream.hpp (hunk 2 above)
Comment (rongou) [reply cpp/include/cudf/utilities/default_stream.hpp:27]: Good catch!

File: cpp/include/cudf/utilities/default_stream.hpp (hunk 1 above)
Comment (rongou) [reply cpp/include/cudf/utilities/default_stream.hpp:52]: Wouldn't making it thread safe introduce a lot of overhead? At least in Spark's use case, we are going to set the default stream once during startup and never change it again. I imagine most use cases would be similar. Seems silly to change the default stream all the time.

The setter only allows `cudaStreamLegacy` or `cudaStreamPerThread`, so multiple devices shouldn't be an issue.

**Comment by rongou** [comment]
Superseded by #6330

**Pull request closed without merging**